starport/pkg/spn/spn.go

```diff
@@ -274,6 +275,34 @@ func (c *Client) broadcast(ctx context.Context, clientCtx client.Context, msgs .
 
 	c.out.Reset()
 
+	// calculate the necessary gas for the transaction
+	txf, err := tx.PrepareFactory(clientCtx, c.factory)
+	if err != nil {
+		return err
+	}
+	c.factory = txf
+	_, gas, err := tx.CalculateGas(clientCtx.QueryWithData, c.factory, msgs...)
+	if err != nil {
+		return nil
+	}
+	// The simulated gas can vary from the actual gas needed for a real transaction
+	// We add an additional amount to endure sufficient gas is provided
+	gas += 10000
+	c.factory = c.factory.WithGas(gas)
+
+	// Prompt for confirmation
+	if confirmPrompt {
```

Comment:
I feel like moving prompt to the `cmd` package.

What about changing `broadcast(...) error` func to more like a provision function `broadcast(...) (gas int64, do func() error, err error)`? 

Once `do()` is executed, it actually broadcasts the tx.

Reply:
@ltacker if you prefer, we can also track this by another issue to improve in a later time.

Reply:
something like this:

in `spn.go`:

```go
func (c *Client) provision(ctx context.Context, clientCtx client.Context, confirmPrompt bool, msgs ...types.Msg) (gas int64, broadcast func() error, err error) {
	// validate msgs.
	for _, msg := range msgs {
		if err := msg.ValidateBasic(); err != nil {
			return 0, nil, err
		}
	}

	// make sure that account has enough balances before broadcasting.
	if err := c.makeSureAccountHasTokens(ctx, clientCtx.GetFromAddress().String()); err != nil {
		return 0, nil, err
	}

	c.out.Reset()

	// calculate the necessary gas for the transaction
	txf, err := tx.PrepareFactory(clientCtx, c.factory)
	if err != nil {
		return 0, nil, err
	}
	c.factory = txf
	_, gas, err := tx.CalculateGas(clientCtx.QueryWithData, c.factory, msgs...)
	if err != nil {
		return 0, nil, err
	}
	// The simulated gas can vary from the actual gas needed for a real transaction
	// We add an additional amount to endure sufficient gas is provided
	gas += 10000
	c.factory = c.factory.WithGas(gas)

	return gas, func() error {
		// broadcast tx.
		if err := tx.BroadcastTx(clientCtx, c.factory, msgs...); err != nil {
			if strings.Contains(err.Error(), "not found") {
				return errors.New("make sure that your SPN account has enough balance")
			}
			return err
		}

		// handle results.
		out := struct {
			Code int    `json:"code"`
			Log  string `json:"raw_log"`
		}{}
		if err := json.NewDecoder(c.out).Decode(&out); err != nil {
			return err
		}
		if out.Code > 0 {
			return fmt.Errorf("SPN error with '%d' code: %s", out.Code, out.Log)
		}
		return nil
	}, nil
}
```

```go
func (c *Client) SubmitReviewals(ctx context.Context, accountName, chainID string, reviewals ...Reviewal) (gas int64, broadcast func() error, err error) {}
````

in `cmd`:
```go
	gas, broadcast, err := nb.SubmitReviewals(cmd.Context(), chainID, reviewals...); 
	if err != nil {
		return err
	}

	// Prompt for confirmation
	if confirmPrompt {
		prompt := promptui.Prompt{
			Label: fmt.Sprintf("This operation will cost about %v gas. Confirm the transaction?",
				gas,
			),
			IsConfirm: true,
		}
		if _, err := prompt.Run(); err != nil {
			return errors.New("transaction aborted")
		}
	}

	if err := broadcast(); err != nil {
		return err
	}
```

Reply:
Thanks for the code, I check that 👍